cms/djangoapps/contentstore/exams.py

```diff
@@ -0,0 +1,123 @@
+from openedx.core.djangoapps.oauth_dispatch.jwt import create_jwt_for_user
+from xmodule.modulestore.django import modulestore
+from xmodule.modulestore.exceptions import ItemNotFoundError
+
+from .views.helpers import is_item_in_course_tree
+
+log = logging.getLogger(__name__)
+User = get_user_model()
+
+
+def register_exams(course_key):
+    """
+    This is typically called on a course published signal. The course is examined for sequences
+    that are marked as timed exams. Then these are registered with the exams service.
+    Likewise, if formerly registered exams are not included in the payload they will
+    be marked inactive by the exam service.
+    """
+    if not settings.FEATURES.get('ENABLE_SPECIAL_EXAMS') or not exams_ida_enabled(course_key):
+        # if feature is not enabled then do a quick exit
+        return
+
+    course = modulestore().get_course(course_key)
+    if course is None:
+        raise ItemNotFoundError("Course {} does not exist", str(course_key))  # lint-amnesty, pylint: disable=raising-format-tuple
+
+    # get all sequences, since they can be marked as timed/proctored exams
+    _timed_exams = modulestore().get_items(
+        course_key,
+        qualifiers={
+            'category': 'sequential',
+        },
+        settings={
+            'is_time_limited': True,
+        }
+    )
+
+    # filter out any potential dangling sequences
+    timed_exams = [
+        timed_exam
+        for timed_exam in _timed_exams
+        if is_item_in_course_tree(timed_exam)
+    ]
+
+    exams_list = []
+    locations = []
+    for timed_exam in timed_exams:
+        location = str(timed_exam.location)
+        msg = (
+            'Found {location} as an exam in course structure.'.format(
+                location=location
+            )
+        )
+        log.info(msg)
+        locations.append(location)
+
+        if timed_exam.is_proctored_exam:
+            if timed_exam.is_onboarding_exam:
+                exam_type = 'onboarding'
+            elif timed_exam.is_practice_exam:
+                exam_type = 'practice_proctored'
+            else:
+                exam_type = 'proctored'
+        else:
+            exam_type = 'timed'
+
+        exams_list.append({
+            'course_id': str(course_key),
+            'content_id': str(timed_exam.location),
+            'exam_name': timed_exam.display_name,
+            'time_limit_mins': timed_exam.default_time_limit_minutes,
+            'due_date': timed_exam.due if not course.self_paced else None,
+            'exam_type': exam_type,
+            'is_active': True,
+            'hide_after_due': timed_exam.hide_after_due,
+            # backend is only required for continued edx-proctoring support
+            'backend': course.proctoring_provider,
+        })
+
+    try:
+        _patch_course_exams(exams_list, str(course_key))
+        log.info(f'Successfully registered {locations} with exam service')
```

Comment:
are the locations useful in this log or too detailed? what about the course key?

Reply:
It doesn't actually come out to messy since I wouldn't expect a large number of exams in one course `Successfully registered ['block-v1:edX+DemoX+Demo_Course+type@sequential+block@00a45f2d4cc54b07a172183a1834d710', 'block-v1:edX+DemoX+Demo_Course+type@sequential+block@00a45f2d4cc54b07a172183a1834d71123'] with exam service`

But you might have a point, I'm not sure if this is actually more useful than just the course key 🤔